### PR TITLE
feat: catalog function to soft delete based on retention

### LINKED ISSUE #6024: Soft delete parquet files that are outside retention period
# Rationale

This is a task of #3053 to create a new background loop in Garbage Collector to find files that outside retention period and mark it to_delete (update `to_delete` with current time).

File that is outside retention period: file's max_time < now - retention_period_nano of the file's namespace

### PR DESCRIPTION
**Pull request by lukebond**
Part of #6024

Adds a function to the parquet file repo in the catalog that will flag files for deletion if they're older than their namespace's retention policy. This will be used in the GC to close #6024.

NOTE: requirements of retention epic #3053 indicate that a table-based retention policy needs to exist. I've left TODO comments where that should be respected, once implemented.

- [ ] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [ ] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: iox_catalog/src/postgres.rs [hunk 1]
```diff
@@ -1646,6 +1646,28 @@ RETURNING *;
         Ok(())
     }
 
+    async fn flag_for_delete_by_retention(&mut self) -> Result<Vec<ParquetFileId>> {
+        let flagged_at = Timestamp::from(self.time_provider.now());
+        // TODO - include check of table retention period once implemented
```
Comment (NGA-TRAN) [inline iox_catalog/src/postgres.rs:1651]: YEah, we can keep this todo but for now, this query to update all files of the namespace is fine

File: iox_catalog/src/interface.rs [hunk 2]
```diff
@@ -2520,6 +2528,70 @@ pub(crate) mod test_helpers {
...
+            .unwrap()
+            .unwrap();
+        assert_matches!(f4.to_delete, Some(_)); // f4 is > 1hr old
+        let f5 = repos
+            .parquet_files()
+            .get_by_object_store_id(f5.object_store_id)
+            .await
+            .unwrap()
+            .unwrap();
+        assert_matches!(f5.to_delete, None); // f5 is < 1hr old
```
Comment (NGA-TRAN) [inline iox_catalog/src/interface.rs:2594]: Nice tests and comments 👍

**Review by NGA-TRAN — APPROVED** [review]
LGTM

File: iox_catalog/src/postgres.rs [hunk 3]
```diff
@@ -1646,6 +1646,28 @@ RETURNING *;
...
+        let flagged = sqlx::query(
+            r#"
+                UPDATE parquet_file
+                SET to_delete = $1
+                FROM namespace
+                WHERE retention_period_ns IS NOT NULL
```
Comment (NGA-TRAN) [inline iox_catalog/src/postgres.rs:1657]: I am working on #6025 and I think you may wan to add `to_delete is null` to only soft delete files that are not yet soft deleted. This way, their to_delete will be older and will be hard deleted sooner

Comment (lukebond) [reply iox_catalog/src/postgres.rs:1657]: have done this in #6073

**Pull request merged**